ops/framework.py

```diff
@@ -1245,7 +1250,12 @@ def _wrapped_repr(obj: '_StoredObject') -> str:
 
 
 class StoredDict(typing.MutableMapping[Hashable, Any]):
-    """A dict-like object that uses the StoredState as backend."""
+    """A dict-like object that uses the StoredState as backend.
+
+    Charms are not expected to use this class directly. Adding a
+    :class:`StoredState` attribute to a charm class will automatically use this
+    class to store dictionary-like data.
```

Comment:
How does Ops know whether to create a StoredDict, StoredList, or StoredSet when you create/access a `_stored` attribute?

Reply:
When you set the attribute [it gets wrapped](https://github.com/canonical/operator/blob/51ba92f06b7b5f2e87c32aa67fb411b583be47ea/ops/framework.py#L1109) and similarly unwrapped when [accessed](https://github.com/canonical/operator/blob/51ba92f06b7b5f2e87c32aa67fb411b583be47ea/ops/framework.py#L1123).

It does an `isinstance` check so I guess the doc should say "dictionary data" rather than 
dictionary-like" even though the `StoredDict` class is more generic.